modules/frontend/querysharding.go

```diff
@@ -63,15 +66,15 @@ func (s shardQuery) Do(r *http.Request) (*http.Response, error) {
 	}
 
 	reqs := make([]*http.Request, s.queryShards)
-	for i := 0; i < len(s.blockBoundaries)-1; i++ {
+	for i := 0; i < s.queryShards; i++ {
```

Comment:
instead of creating block boundaries, it would be a lot cleaner to make a method that created a slice of structs that represented the query.  like:
```
type queryParams struct {
  queryShards bool
  startBlock []byte
  endBlock []byte
}
```
that would simplify this logic quite a bit and would be much easier to test as well.